### PR TITLE
add byte serializer library

### PR DESCRIPTION
**Pull request by bogdan34**

### CODE REVIEW

File: shared/util/byteSerializer/config.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
+import { BufferTypes } from 'shared/util/byteSerializer/types';
+import { schemaStore } from 'shared/util/byteSerializer/schema';
+
+export enum SchemaName {
+    request = 0,
```
Comment (macaqueee) [inline shared/util/byteSerializer/config.ts]: CAPITAL

File: shared/util/byteSerializer/config.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    Empty = 16
+}
+
+schemaStore.add(SchemaName.Request, {
+    code: new BufferTypes.Utf8(),
+    data: new BufferTypes.Buffer(),
+});
+
+schemaStore.add(SchemaName.ShortPeerInfo, {
+    ip: new BufferTypes.Utf8(),
```
Comment (Verbalist) [inline shared/util/byteSerializer/config.ts]: TODO can be stored in 4 byte

File: shared/util/byteSerializer/config.ts [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    broadhash: new BufferTypes.Utf8(),
+    peerCount: new BufferTypes.Uint16()
+});
+
+schemaStore.add(SchemaName.FullHeaders, {
+    height: new BufferTypes.Uint32(),
+    broadhash: new BufferTypes.Utf8(),
+    blocksIds: new BufferTypes.Map(new BufferTypes.Uint32(), new BufferTypes.Utf8()),
+    os: new BufferTypes.Utf8(),
+    version: new BufferTypes.Utf8(),
```
Comment (Verbalist) [inline shared/util/byteSerializer/config.ts]: TODO can be stored in 3 byte

File: shared/util/byteSerializer/config.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    minVersion: new BufferTypes.Utf8(),
```
Comment (Verbalist) [inline shared/util/byteSerializer/config.ts]: TODO can be stored in 3 byte

File: shared/util/byteSerializer/config.ts [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    generatorPublicKey: new BufferTypes.Utf8(),
+    signature: new BufferTypes.Utf8(),
+    relay: new BufferTypes.Uint8(),
+    transactions: new BufferTypes.Buffer(),
+});
+
+schemaStore.add(SchemaName.Transaction, {
+    id: new BufferTypes.Utf8(),
+    blockId: new BufferTypes.Utf8(),
+    type: new BufferTypes.Uint16(),
```
Comment (Verbalist) [inline shared/util/byteSerializer/config.ts]: UInt8

File: shared/util/byteSerializer/models/map/simpleArray.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        this.typeElement = type;
+    }
+
+    getLength(value: Array<any>) {
+        return this.length.head + this.getBodyLength(value);
+    }
+
+    private getBodyLength(value: Array<any>): number {
+        let size = 0;
+        value.forEach(item => {
```
Comment (macaqueee) [inline shared/util/byteSerializer/models/map/simpleArray.ts]: reduce

File: shared/util/byteSerializer/models/array.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    }
+
+    getLength(value: Array<any>): number {
+        return this.length.type + this.length.type + this.length.head + this.getBodyLength(value);
+    }
+
+    private getBodyLength(value: Array<any>): number {
+        let size = 0;
+        value.forEach(item => {
+            size += this.typeElement.getLength(item);
```
Comment (macaqueee) [inline shared/util/byteSerializer/models/array.ts]: reduce

**Pull request merged**